src/ngraph_assign_clusters.cc

```diff
@@ -493,14 +543,46 @@ Status AssignClusters(Graph* graph) {
         MergeClusters(edge, cluster_map);
         // something changed
         changed = true;
+      } else {
+        if (collect_non_contracting_edge_info) {
+          // either static input
+          // or there exists a longer path, so contracting this edge causes
+          // cycles
+          std::vector<int32> static_inputs;
+          GetStaticInputs(dst, &static_inputs);
+          bool is_static = std::find(static_inputs.begin(), static_inputs.end(),
+                                     edge->dst_input()) != static_inputs.end();
+          if (is_static) {
+            bool fed_by_const = src->type_string() == "Const";
+            if (fed_by_const) {
+              errors::Internal("A Const node ", src->name(),
+                               " is feeding a static input in ", dst->name(),
+                               ", but they are in different clusters. This "
+                               "should not be happening");
+            }
+          }
+          cluster_separation_reason[get_string_key(src_index, dst_index)]
+              .push_back(is_static ? EdgeNonContractionReasons::STATICINPUT
+                                   : EdgeNonContractionReasons::PATHEXISTS);
+        }
       }
     }
+
+    if (!changed && config::IsLoggingPlacement() &&
+        !collect_non_contracting_edge_info) {
+      changed = true;
+      collect_non_contracting_edge_info = true;
+    }
+    if (collect_non_contracting_edge_info) {
```

Comment:
Is this `if` statement doing anything here?  `collect_non_contracting_edge_info`  seems only = true when the above if looped is entered at the end of the clustering. I guess you could add an assert statement here for `changed` variable.

Reply:
Ah yes, had a comment there, forgot to add it later :-D